blog-posts/common-approaches-to-handling-subscriptions/common-approaches-to-handling-subscriptions.md

```diff
@@ -0,0 +1,178 @@
+---
+published: true
+title: "Common Approaches to Handling Subscriptions"
+cover_image: ""
+description: "This blog article covers the some common approaches to handling Observable subscriptions with RxJS."
+tags: rxjs, javascript, typescript, angular
+series:
+canonical_url:
+---
+
+When developers first start using RxJS one of the biggest challenges is handling subscriptions.
+
+**Subscriptions** in RxJS handle the execution flow of a stream. When we create Observables, we "subscribe" to them to start using them. Conversely, when we "unsubscribe" we do the reverse to stop the same execution flow.
+
+Handling this can be a bit tricky. This post is going to cover some common patterns for handling subscriptions in your code.
+
+This post is also going to be framework agnostic in an effort to make these patterns accessible by all.
+
+The examples that are used in this post can be reached in my [Stackblitz Project](https://rxjs-example-handle-susbcriptions.stackblitz.io).
+
+I'm going to show the code here, and have an embedded link to my Stackblitz project at the end. I encourage you to run the code examples that I walkthrough to get a better understanding.
+
+## Memory Leaks and Your First Unsubscribe
+
+When we do not successfully unsubscribe from an Observable, we create a situation called a "memory leak." This is any time a stream is started (using system resources) and not stopped. If you have enough streams started without an "unsubscribe," you can use up a lot of your system resources and eventually bring down your session..._this is not a good thing_.
+
+A good example of this would be a simple Observable from the creation operator `interval`. Consider the following code:
+
+```javascript
+import { interval } from 'rxjs';
+
+const observable = interval(1000);
+const subscription = observable.subscribe(() => console.log('Hello!'));
+```
+
+So in this example we are just using the `interval` operator to create a stream that writes "Hello!" to the console every 1 second. When we call `subscribe` we are saying that whenever the stream emits a response (in this case every 1 second), we print "Hello!".
+
+This is very simplistic, but the challenge here is that if we do not call `unsubscribe`, this stream will continue running until you end your session or destroy the associated component etc. This is really easy to miss and important for performance.
+
+To fix this situation, a simple "unsubscribe" is needed. So consider the same code but with the addition of an "unsubscribe" call like so:
+
+```javascript
+import { interval } from 'rxjs';
+
+const observable = interval(1000);
+const subscription = observable.subscribe(() => console.log('Hello!'));
+subscription.unsubscribe();
```

Comment:
maybe you should add a `setTimeout` here because this won't print anything to the console

Reply:
Agree, I went ahead and added a "unsubscribed" message to the console output as well with a note.

```javascript
import { interval } from 'rxjs';

const observable = interval(1000);
const subscription = observable.subscribe(() => console.log('Hello!'));
setTimeout(() => {
  subscription.unsubscribe();
  console.log('unsubscribed');
}, 1000);
```